src/Database/LSMTree/Internal.hs

```diff
@@ -558,16 +558,198 @@ updates ::
+
+{-# SPECIALISE updatesWithInterleavedFlushes :: TableConfig -> HasFS IO h -> SessionRoot -> UniqCounter IO -> V.Vector (SerialisedKey, Entry SerialisedValue SerialisedBlob) -> TempRegistry IO (Run (Handle h)) -> TableContent h -> IO (TableContent h) #-}
+-- | A single batch of updates can fill up the write buffer multiple times. We
+-- flush the write buffer each time it fills up before trying to fill it up
+-- again.
+updatesWithInterleavedFlushes ::
+     m ~ IO
+  => TableConfig
+  -> HasFS m h
+  -> SessionRoot
+  -> UniqCounter m
+  -> V.Vector (SerialisedKey, Entry SerialisedValue SerialisedBlob)
+  -> TempRegistry m (Run (Handle h))
+  -> TableContent h
+  -> m (TableContent h)
+updatesWithInterleavedFlushes conf hfs root uniqC es0 reg tc = do
+    case updatesWriteBuffer conf es0 (tableWriteBuffer tc) of
+      NotYetFull wb' -> pure $ setWriteBuffer wb' tc
+      Full wb' es' -> do
+        let tc' = setWriteBuffer wb' tc
+        tc'' <- flushWriteBuffer hfs root uniqC reg tc'
+        updatesWithInterleavedFlushes conf hfs root uniqC es' reg tc''
+  where
+    setWriteBuffer :: WriteBuffer -> TableContent h -> TableContent h
+    setWriteBuffer wb' tc0 = TableContent {
+          tableWriteBuffer = wb'
+        , tableLevels = tableLevels tc0
+        , tableCache = tableCache tc0
+        }
+
+-- | The result of a batch of entries to a write buffer.
+data UpdateResult =
+    NotYetFull !WriteBuffer
+  | Full !WriteBuffer !(V.Vector (SerialisedKey, Entry SerialisedValue SerialisedBlob))
+
+-- | Add entries to a write buffer until we run out of entries ('NotYetFull'),
+-- or until the write buffer becomes full ('Full'),
+--
+-- TODO: in case the write buffer is empty, or if it fills up multiple times for
+-- a single batch of updates, we might be able to skip adding entries to the
+-- write buffer for a large part. We can sort and deduplicate the vector of
+-- entries directly, slice it up into properly sized sub-vectors, and write
+-- those to disk.
+--
+-- TODO: should we just flush the write buffer it turns out the batch is so
+-- large that it /might/ fill up the write buffer? If we do so, then we can take
+-- the approach from the previous TODO to flush vectors to disk directly.
+updatesWriteBuffer ::
+     TableConfig
+  -> V.Vector (SerialisedKey, Entry SerialisedValue SerialisedBlob)
+  -> WriteBuffer
+  -> UpdateResult
+updatesWriteBuffer conf es0 wb0 = go es0 wb0
   where
-    resolve = resolveMupsert (tableConfig th)
+    AllocNumEntries (NumEntries maxn) = confWriteBufferAlloc conf
+    resolve = resolveMupsert conf
+
+    -- A single update is not guaranteed to increase the size of the write
+    -- buffer: either the size stays the same because a key already exists, or
+    -- the size increases by 1 because a key did /not/ already exist. Because of
+    -- this, we recursively add entries to the write buffer until it is filled
+    -- up.
+    --
+    -- TODO: can we update the write buffer non-recursively?
```

Comment:
This is fine, but when we spend more time on optimisation we might consider:

* noting that we can allowing a range of sizes to flush to disk rather than just the max size
* could do a map bulk merge rather than sequential insert, on the prefix of the batch that's guaranteed to fit
* or flush the existing buffer if we would expect the next batch to cause the buffer to become too large

Reply:
Map merge is probably indeed faster than a loop of individual inserts (but might allocate a little extra?).

> TODO: can we update the write buffer non-recursively?

I think some recursion is necessary here, e.g. if a batch is much larger than the remaining space in the buffer, but shares all its keys with the buffer, you kind of have to go through all the keys in small chunks. The number of iterations needed gets much smaller when there is a range of sizes that are okay to flush (as suggested), but the issue remains.

Reply:
I've added the suggestions to the comment.